app/components/root/root.js

```diff
@@ -29,21 +40,98 @@ export default class Root extends React.Component {
     componentDidMount() {
         AppState.addEventListener('change', this.handleAppStateChange);
         EventEmitter.on(Constants.CONFIG_CHANGED, this.handleConfigChanged);
+        MessageBarManager.registerMessageBar(this.refs.notification);
+        this.configurePushNotifications();
     }
 
     componentWillUnmount() {
         AppState.removeEventListener('change', this.handleAppStateChange);
         EventEmitter.off(Constants.CONFIG_CHANGED, this.handleConfigChanged);
+        MessageBarManager.unregisterMessageBar();
+        PushNotification.unregister();
     }
 
+    onRegisterDevice = (data) => {
+        const prefix = Platform.OS === 'ios' ? Constants.PUSH_NOTIFY_APPLE_REACT_NATIVE : Constants.PUSH_NOTIFY_ANDROID_REACT_NATIVE;
+        this.props.actions.setDeviceToken(`${prefix}:${data.token}`);
+    };
+
+    onPushNotification = (notification) => {
+        const {foreground, userInteraction, data, message} = notification;
+        let noty;
+
+        if (Platform.OS === 'android') {
+            noty = {
+                data: {
+                    channel_id: notification.channel_id,
+                    team_id: notification.team_id
+                },
+                message
+            };
+        } else {
+            noty = {
+                data,
+                message
+            };
+        }
+
+        if (foreground) {
+            this.handleInAppNotification(noty);
+        } else if (userInteraction) {
+            this.onNotificationTapped(noty);
+        }
+    };
+
+    onNotificationTapped = (notification) => {
+        const {currentTeamId, currentChannelId} = this.props;
+
+        if (currentTeamId && currentChannelId) {
+            // this means that the store has the necessary data
+            this.props.actions.goToNotification(notification);
+        } else {
+            this.props.actions.queueNotification(notification);
+        }
+    };
+
     handleAppStateChange(appState) {
         this.props.actions.setAppState(appState === 'active');
+
+        if (appState === 'inactive') {
+            this.props.actions.flushToStorage();
```

Comment:
Nothing to change for this right now, but I came across something where they mentioned using `store.subscribe` to listen for store changes and automatically write those into storage in the background so that it's automatically managed. That might be worth looking into at some point.